src/App.tsx

```diff
@@ -1,223 +1,220 @@
+    }
+
+    loadTodos(user.id);
+  }, [user]);
 
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
     if (newTodoField.current) {
       newTodoField.current.focus();
     }
-  }, []);
+  }, [todos]);
+
+  const visibleTodos = filterType === FilterType.All
+    ? todos
+    : todos.filter(({ completed }) => {
+      switch (filterType) {
+        case FilterType.Active:
+          return !completed;
+        case FilterType.Completed:
+          return completed;
+
+        default:
+          throw new Error();
+      }
+    });
+
+  const handleChooseFilter = useCallback(
+    (filter: FilterType) => {
+      setFilterType(filter);
+    },
+    [filterType],
+  );
+
+  const handleSubmit = useCallback(async (event: FormEvent) => {
+    event.preventDefault();
+
+    setIsAdding(true);
+
+    if (!title.trim()) {
+      setTodosError(TodosError.Title);
+      setTitle('');
+
+      return;
+    }
+
+    try {
+      if (!user) {
+        return;
+      }
+
+      const newTodo = await addTodos(user.id, title);
+
+      setTodos([...todos, newTodo]);
+    } catch {
+      setTodosError(TodosError.Adding);
+    }
+
+    setTitle('');
+    setIsAdding(false);
+  }, [title, user]);
+
+  const handleDelete = useCallback(async (todoId: number) => {
+    try {
+      await deleteTodo(todoId);
+
+      setTodos([...todos.filter(({ id }) => id !== todoId)]);
+    } catch {
+      setTodosError(TodosError.Deleting);
+    }
+  }, [todos, todosError, isAdding]);
+
+  const handleChangeInput = ({
+    target: { value },
+  }: React.ChangeEvent<HTMLInputElement>) => setTitle(value);
+
+  const toggleAllActive = todos.every(({ completed }) => completed);
+
+  const upgradeTodos = useCallback(
+    async (todoId: number, data: Partial<Todo>) => {
+      try {
+        const patchedTodo: Todo = await patchTodo(todoId, data);
+
+        setTodos(todos.map(todo => (
+          todo.id === todoId
+            ? patchedTodo
+            : todo
+        )));
+      } catch {
+        setTodosError(TodosError.Updating);
+      }
+    }, [todos],
+  );
+
+  const uncompletedTodos = todos.filter(({ completed }) => !completed);
+
+  const handleClickToggle = () => {
+    if (uncompletedTodos.length) {
+      uncompletedTodos.map(({ id }) => patchTodo(id,
+        { completed: true }).catch(() => setTodosError(TodosError.Updating)));
+      setTodos(todos.map(todo => {
+        const copy = todo;
+
+        copy.completed = true;
+
+        return copy;
+      }));
+    } else {
+      todos.map(({ id }) => patchTodo(id,
+        { completed: false }).catch(() => setTodosError(TodosError.Updating)));
+      setTodos(todos.map(todo => {
+        const copy = todo;
+
+        copy.completed = false;
+
+        return copy;
+      }));
+    }
+  };
+
+  const handleClearCompleted = () => {
+    todos.forEach(({ id, completed }) => {
+      if (completed) {
+        deleteTodo(id)
+          .catch(() => setTodosError(TodosError.Deleting));
+      }
```

Comment:
Try to use  [promise all + map](https://betterprogramming.pub/how-to-use-async-await-with-map-in-js-5059043564e0)